FaunaDB.Client/Types/Decoder.cs

```diff
@@ -0,0 +1,469 @@
+                        return FromArray(value, dstType);
+
+                    if (typeof(IDictionary).IsAssignableFrom(dstType) || (dstType.IsGenericType && dstType.Is(typeof(IDictionary<,>))))
+                        return FromDictionary(value, dstType);
+
+                    if (typeof(IList).IsAssignableFrom(dstType) || (dstType.IsGenericType && dstType.Is(typeof(IEnumerable<>))))
+                        return FromEnumerable(value, dstType);
+
+                    return FromObject(value, dstType);
+            }
+
+            throw new InvalidOperationException($"Cannot convert {value} to {dstType}");
+        }
+
+        static IList FromEnumerable(Value value, Type type)
+        {
+            if (!(value is ArrayV))
+                throw new InvalidOperationException($"Cannot convert `{value}` to {type}");
+
+            if (!type.IsGenericType)
+                throw new InvalidOperationException($"The type {type} is not generic");
+
+            var valueType = type.GenericTypeArguments[0];
+
+            var createType = type.IsAbstract
+                                 ? typeof(List<>).MakeGenericType(valueType)
+                                 : type;
+
+            var ret = (IList)Activator.CreateInstance(createType);
+
+            foreach (var v in ((ArrayV)value).Value)
+                ret.Add(Decode(v, valueType));
+
+            return ret;
+        }
+
+        static IDictionary FromDictionary(Value value, Type type)
+        {
+            if (!(value is ObjectV))
+                throw new InvalidOperationException($"Cannot convert `{value}` to {type}");
+
+            if (!type.IsGenericType)
+                throw new InvalidOperationException($"The type {type} is not generic");
+
+            var keyType = type.GenericTypeArguments[0];
+            var valueType = type.GenericTypeArguments[1];
+
+            var createType = type.IsAbstract
+                                 ? typeof(Dictionary<,>).MakeGenericType(keyType, valueType)
+                                 : type;
+
+            var ret = (IDictionary)Activator.CreateInstance(createType);
+
+            foreach (var kv in ((ObjectV)value).Value)
+                ret.Add(kv.Key, Decode(kv.Value, valueType));
+
+            return ret;
+        }
+
+        static Array FromArray(Value value, Type type)
+        {
+            var elementType = type.GetElementType();
+
+            if (!(value is ArrayV))
+                throw new InvalidOperationException($"Cannot convert `{value}` to an array of type {elementType}");
+
+            var array = ((ArrayV)value).Value;
+
+            var ret = Array.CreateInstance(elementType, array.Count);
+
+            for (var i = 0; i < array.Count; i++)
+                ret.SetValue(Decode(array[i], elementType), i);
+
+            return ret;
+        }
+
+        static T FromScalar<T, R>(Value value)
+        {
+            if (value is ScalarValue<T>)
+                return ((ScalarValue<T>)value).Value;
```

Comment:
So `R` is only used to build the error message, right? Maybe you can compose with another function to build the error message to avoid the type information in the function signature. It's a little confusing.